Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 4d03051016f7d682114c3e0a01d383e5b2dc3eaf
Number of Error(s): 0
CMake Error at cmake_modules/commonSetup.cmake:480 (add_library):
Number of Error(s): 0
CMakeFiles/CMakeOutput.log copied into ../CMakeOutput-2019-03-15_03-58-44.log
CMakeFiles/CMakeError.log copied into ../CMakeError-2019-03-15_03-58-44.log

[Build error:](https://gist.github.com/93c596be2cb6ce9603f9cda0ec2a012c)
Build log [Build log](https://gist.github.com/HPCCSmoketest/93c596be2cb6ce9603f9cda0ec2a012c#file-result-19-03-15-03-58-24-log)

--- a/esp/services/ws_ecl/ws_ecl_service.cpp
+++ b/esp/services/ws_ecl/ws_ecl_service.cpp
@@ -1899,9 +1899,8 @@ int CWsEclBinding::submitWsEclWorkunit(IEspContext & context, WsEclWuInfo &wsinf
             workunit->setDebugValue("GlobalId", globalId.str(), true);
             workunit->setDebugValue("GlobalIdHeader", globalIdHeader.str(), true);  //use same header received
 
-            SocketEndpoint ep;
             StringBuffer localId;
-            appendLocalId(localId, httpreq->getSocket()->getEndpoint(ep), 0);
+            appendGloballyUniqueId(localId);
             workunit->setDebugValue("CallerId", localId.str(), true); //our localId becomes caller id for the next hop
             workunit->setDebugValue("CallerIdHeader", callerIdHeader.str(), true); //use same header received
             DBGLOG("GlobalId: %s, CallerId: %s, LocalId: %s, Wuid: %s", globalId.str(), callerId.str(), localId.str(), wuid.str());
@@ -1998,9 +1997,8 @@ void CWsEclBinding::sendRoxieRequest(const char *target, StringBuffer &req, Stri
                 headers.setown(createProperties());
                 headers->setProp(globalIdHeader, globalId);
 
-                SocketEndpoint ep;
                 StringBuffer localId;
-                appendLocalId(localId, httpreq->getSocket()->getEndpoint(ep), 0);
+                appendGloballyUniqueId(localId);
                 if (localId.length())
                     headers->setProp(callerIdHeader, localId);
                 DBGLOG("GlobalId: %s, CallerId: %s, LocalId: %s", globalId.str(), callerId.str(), localId.str());

--- a/roxie/ccd/ccd.hpp
+++ b/roxie/ccd/ccd.hpp
@@ -617,7 +617,7 @@ public:
     virtual void setGlobalId(const char *id, SocketEndpoint &ep, unsigned pid) override
     {
         globalId.set(id);
-        appendLocalId(localId.clear(), ep, pid);
+        appendGloballyUniqueId(localId.clear());
     }
     virtual void setCallerId(const char *id) override
     {

--- a/roxie/ccd/ccdlistener.cpp
+++ b/roxie/ccd/ccdlistener.cpp
@@ -1393,6 +1393,13 @@ public:
             throw MakeStringException(ROXIE_NO_PACKAGES_ACTIVE, "Unknown query %s (no packages active)", queryName.get());
         }
         queryFactory->checkSuspended();
+        IConstWorkUnit *workunit = queryFactory->queryWorkUnit();
+        if (workunit && workunit->getDebugValueBool("generateGlobalId", false) && isEmptyString(logctx->queryGlobalId()))
+        {
+            StringBuffer gen_id;
+            appendGloballyUniqueId(gen_id);
+            setTransactionId(gen_id.str(), nullptr, true);
+        }
     }
     virtual void noteQueryActive()
     {

--- a/system/globalid/lnuid.cpp
+++ b/system/globalid/lnuid.cpp
@@ -1,0 +1,384 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2019 HPCC Systems®.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#ifdef __linux__
+#include <sys/time.h>
+#elif _WIN32
+#include <windows.h>
+#include <wincrypt.h>
+#include <chrono>
+#else
+
+#endif
+
+/**
+* This is an implementation of Globally Unique Transaction ID’s.
+* Note: This class currently generates unique id on OS's that supports device /dev/urandom.
+* @author - Amol Patwardhan
+* Version: 1.0
+*/
+
+#include <iostream>
+#include <stdio.h>
+#include <string.h>
+#include <vector>
+#include <assert.h>
+#include <stdio.h>
+#include <time.h>
+#include <sstream>
+#include <ctime>
+#include "lnuid.h"
+using namespace std;
+
+namespace ln_uid {
+
+    /** All alphanumeric characters except for "0", "I", "O", and "l" */
+    const char* pszBase58 = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
+
+    //Random byte count
+    const unsigned int random_byte_count = 11;
+    //Random byte count
+    const unsigned int time_byte_count = 5;
+
+    //NOTE: This variable is used to offset dropped 41st bit from time component.
+    //      This variable adds time till last 41st bit flip on 03-Nov-2004.
+    //      This value should be updated later to handle next bit flip on 07-Sep-2039.
+    unsigned long long  int OFFSET = (unsigned long long  int) 1099511627776;
+
+
+    ln_uid_t &createUniqueId(ln_uid_t &out)
+    {
+#ifdef __linux__
+        struct timeval tv;
+        gettimeofday(&tv, NULL);
+
+        unsigned long long  int millisecondsSinceEpochTrim =
+            (unsigned long long int)(tv.tv_sec) * 1000 +
+            (unsigned long long int)(tv.tv_usec) / 1000;
+
+#elif _WIN32
+        unsigned long long  int millisecondsSinceEpochTrim =
+            std::chrono::system_clock::now().time_since_epoch() / std::chrono::milliseconds(1);
+#else
+
+#endif
+        unsigned char timedata[time_byte_count];
+
+        timedata[0] = (int)((millisecondsSinceEpochTrim >> 32) & 0xFF);
+        timedata[1] = (int)((millisecondsSinceEpochTrim >> 24) & 0xFF);
+        timedata[2] = (int)((millisecondsSinceEpochTrim >> 16) & 0xFF);
+        timedata[3] = (int)((millisecondsSinceEpochTrim >> 8) & 0XFF);
+        timedata[4] = (int)((millisecondsSinceEpochTrim & 0XFF));
+
+        unsigned char randomdata[random_byte_count];
+
+#ifdef __linux__
+        FILE *fp;
+        fp = fopen("/dev/urandom", "r");
+        fread(&randomdata, 1, random_byte_count, fp);
+        fclose(fp);
+#elif _WIN32
+        HCRYPTPROV hProvider;
+
+        CryptAcquireContext(&hProvider, 0, 0, PROV_RSA_FULL, CRYPT_VERIFYCONTEXT);
+        CryptGenRandom(hProvider, random_byte_count, randomdata);
+#endif
+
+        for(unsigned int i=0; i <time_byte_count; i++)
+            out[i] = timedata[i];
+
+        for(unsigned int i=5, j=0; i <uid_size; i++, j++)
+            out[i] = randomdata[j];
+
+        return out;
+    }
+
+    string createUniqueIdString()
+    {
+        ln_uid_t ln_uid = {};
+        createUniqueId(ln_uid);
+        return uniqueIdToString(ln_uid);
+    }
+
+    string uniqueIdToString(const ln_uid_t &uid)
+    {
+        // Skip and count leading zeroes.
+        int zeroes = 0;
+        int length = 0;
+        int counter = 0;
+        while (counter != uid_size && uid[counter] == 0)
+        {
+            counter++;
+            zeroes++;
+        }
+        // Allocate enough space in big-endian base58 representation.
+        int size = (uid_size) * 138 / 100 + 1; // log(256) / log(58), rounded up.
+        vector<unsigned char> b58(size);
+
+        // Process the bytes.
+        while (counter != uid_size)
+        {
+            int carry = uid[counter];
+            int i = 0;
+            // Apply "b58 = b58 * 256 + ch".
+            for (vector<unsigned char>::reverse_iterator it = b58.rbegin(); (carry != 0 || i < length) && (it != b58.rend()); it++, i++)
+            {
+                carry += 256 * (*it);
+                *it = carry % 58;
+                carry /= 58;
+            }
+
+            assert(carry == 0);
+            length = i;
+            counter++;
+        }
+
+        // Skip leading zeroes in base58 result.
+        vector<unsigned char>::iterator it = b58.begin() + (size - length);
+        while (it != b58.end() && *it == 0)
+            it++;
+
+        // Translate the result into a string.
+        string str;
+        str.reserve(zeroes + (b58.end() - it));
+        str.assign(zeroes, '1');
+        while (it != b58.end())
+            str += pszBase58[*(it++)];
+        return str;
+    }
+
+    ln_uid_t &uniqueIdFromString(const char* uid, ln_uid_t &out)
+    {
+        // Skip leading spaces. psz
+        while (*uid && isspace(*uid))
+            uid++;
+        // Skip and count leading '1's.
+        int zeroes = 0;
+        while (*uid == '1')
+        {
+            zeroes++;
+            uid++;
+        }
+        // Allocate enough space in big-endian base256 representation.
+        vector<unsigned char> b256(strlen(uid) * 733 / 1000 + 1); // log(58) / log(256), rounded up.
+
+        // Process the characters.
+        while (*uid && !isspace(*uid))
+        {
+            // Decode base58 character
+            const char* ch = strchr(pszBase58, *uid);
+            if (ch == NULL)
+               return out;
+
+            // Apply "b256 = b256 * 58 + ch".
+            int carry = ch - pszBase58;
+            for (vector<unsigned char>::reverse_iterator it = b256.rbegin(); it != b256.rend(); it++)
+            {
+                carry += 58 * (*it);
+                *it = carry % 256;
+                carry /= 256;
+            }
+            assert(carry == 0);
+            uid++;
+        }
+
+        // Skip trailing spaces.
+        while (isspace(*uid))
+            uid++;
+        if (*uid != 0)
+            return out;
+        // Skip leading zeroes in b256.
+        vector<unsigned char>::iterator it = b256.begin();
+        while (it != b256.end() && *it == 0)
+            it++;
+        vector<unsigned char> vch;
+
+        // Copy result into output vector.
+        vch.reserve(zeroes + (b256.end() - it));
+        vch.assign(zeroes, 0x00);
+        while (it != b256.end())
+            vch.push_back(*(it++));
+
+        for(unsigned int i=0; i <vch.size(); i++)
+            out[i] = vch[i];
+
+        return out;
+    }
+
+    time_t timeFromUniqueId(const char* uid)
+    {
+        ln_uid_t uid_ch = {};
+        uniqueIdFromString(uid, uid_ch);
+        return timeFromUniqueId(uid_ch);
+    }
+
+    time_t timeFromUniqueId(const ln_uid_t &uid)
+    {
+
+        unsigned long long  int timepart = 0;
+        for(unsigned int i=0; i< time_byte_count; i++ )
+            timepart = (timepart << 8) + (uid[i] & 0xff);
+
+        timepart = timepart + OFFSET;
+        time_t seconds = (time_t)(timepart/1000);
+
+        return seconds;
+    }
+
+    void getUniqueIdRange(time_t start, time_t end, ln_uid_t &uid_start, ln_uid_t &uid_end)
+    {
+        unsigned long long int startMilliSecondsSinceEpochTrim = (unsigned long long int)(start) * 1000;
+        unsigned long long int endMilliSecondsSinceEpochTrim = (unsigned long long int)(end) * 1000;
+
+        unsigned char startTimeData[time_byte_count], endTimeData[time_byte_count];
+
+        startTimeData[0] = (int)((startMilliSecondsSinceEpochTrim >> 32) & 0xFF) ;
+        startTimeData[1] = (int)((startMilliSecondsSinceEpochTrim >> 24) & 0xFF) ;
+        startTimeData[2] = (int)((startMilliSecondsSinceEpochTrim >> 16) & 0xFF) ;
+        startTimeData[3] = (int)((startMilliSecondsSinceEpochTrim >> 8) & 0XFF);
+        startTimeData[4] = (int)((startMilliSecondsSinceEpochTrim & 0XFF));
+
+        endTimeData[0] = (int)((endMilliSecondsSinceEpochTrim >> 32) & 0xFF) ;
+        endTimeData[1] = (int)((endMilliSecondsSinceEpochTrim >> 24) & 0xFF) ;
+        endTimeData[2] = (int)((endMilliSecondsSinceEpochTrim >> 16) & 0xFF) ;
+        endTimeData[3] = (int)((endMilliSecondsSinceEpochTrim >> 8) & 0XFF);
+        endTimeData[4] = (int)((endMilliSecondsSinceEpochTrim & 0XFF));
+
+        unsigned char blankedRandomdata[random_byte_count], filledRandomData[random_byte_count];
+
+        for(unsigned int i=0; i <random_byte_count; i++)
+        {
+            blankedRandomdata[i] = 0x00;
+            filledRandomData[i] = 0xFF;
+        }
+
+        for(unsigned int i=0; i <time_byte_count; i++)
+            uid_start[i] = startTimeData[i] ;
+
+        for(unsigned int i=time_byte_count, j=0; i <uid_size; i++, j++)
+            uid_start[i] = blankedRandomdata[j] ;
+
+        for(unsigned int i=0; i <time_byte_count; i++)
+            uid_end[i] = endTimeData[i] ;
+
+        for(unsigned int i=time_byte_count, j=0; i <uid_size; i++, j++)
+            uid_end[i] = filledRandomData[j] ;
+    }
+
+    void getUniqueIdDateRange(const char *start, const char *end, ln_uid_t &uid_start, ln_uid_t &uid_end)
+    {
+        tm startDateTime;
+        tm endDateTime;
+
+        time_t t1, t2, start_time, end_time;
+        t1 = time(NULL);
+        t2 = time(NULL);
+
+#ifdef __linux__
+        localtime_r(&t1, &startDateTime);
+        strptime(start, "%Y-%m-%d %H:%M:%S", &startDateTime);
+
+#elif _WIN32
+        int Y1, m1, d1, H1, M1, S1;
+
+        localtime_s( &startDateTime, &t1);
+        sscanf_s(start, "%4d-%2d-%2d %2d:%2d:%2d", &Y1, &m1, &d1, &H1, &M1, &S1);
+        startDateTime.tm_year = Y1 - 1900;
+        startDateTime.tm_mon = m1 - 1;
+        startDateTime.tm_mday = d1;
+        startDateTime.tm_hour = H1;
+        startDateTime.tm_min = M1;
+        startDateTime.tm_sec = S1;
+#endif
+
+#ifdef __linux__
+        localtime_r(&t2, &endDateTime);
+        strptime(end, "%Y-%m-%d %H:%M:%S", &endDateTime);
+
+#elif _WIN32
+        localtime_s(&endDateTime, &t2);
+        int Y2, m2, d2, H2, M2, S2;
+        sscanf_s(end, "%4d-%2d-%2d %2d:%2d:%2d", &Y2, &m2, &d2, &H2, &M2, &S2);
+        endDateTime.tm_year = Y2 - 1900;
+        endDateTime.tm_mon = m2 - 1;
+        endDateTime.tm_mday = d2;
+        endDateTime.tm_hour = H2;
+        endDateTime.tm_min = M2;
+        endDateTime.tm_sec = S2;
+#endif
+
+        start_time = mktime(&startDateTime) ;
+        end_time = mktime(&endDateTime) ;
+        cout << "Start time:" << "UTC OFFSET:"<< get_utc_offset () << ":" <<start << ":" << start_time << endl;
+        cout << "End time:" << end << ":" << end_time << endl;
+
+        struct tm tm;
+        char sb[100];
+#ifdef __linux__
+        localtime_r(&start_time, &tm);
+        sprintf(sb, "now: %4d-%02d-%02d %02d:%02d:%02d\n", tm.tm_year + 1900, tm.tm_mon + 1, tm.tm_mday,
+            tm.tm_hour, tm.tm_min, tm.tm_sec);
+#elif _WIN32
+        localtime_s(&tm, &start_time);
+        sprintf_s(sb, "now: %4d-%02d-%02d %02d:%02d:%02d\n", tm.tm_year + 1900, tm.tm_mon + 1, tm.tm_mday,
+            tm.tm_hour, tm.tm_min, tm.tm_sec);
+#else
+
+#endif
+        cout << "DateTime is " << sb << endl;
+
+        getUniqueIdRange(start_time, end_time, uid_start, uid_end);
+
+    }
+
+    bool sameUniqueId(const ln_uid_t &uid1, const ln_uid_t &uid2)
+    {
+        return memcmp(uid1, uid2, sizeof uid2)==0;
+    }
+
+    void copyUniqueId(ln_uid_t &to, const ln_uid_t &from)
+    {
+        memcpy (to, from, sizeof from);
+    }
+
+
+/*  returns the utc timezone offset (e.g. -8 hours for PST)
+*/
+    int get_utc_offset()
+    {
+        time_t zero = 24 * 60 * 60L;
+        struct tm timeptr;
+        int gmtime_hours;
+
+        /* get the local time for Jan 2, 1900 00:00 UTC */
+#ifdef __linux__
+        localtime_r(&zero, &timeptr);
+#elif _WIN32
+        localtime_s(&timeptr, &zero);
+#endif
+
+        gmtime_hours = timeptr.tm_hour;
+
+/* if the local time is the "day before" the UTC, subtract 24 hours
+   from the hours to get the UTC offset
+*/
+        if (timeptr.tm_mday < 2)
+            gmtime_hours -= 24;
+
+        return gmtime_hours;
+    }
+};

--- a/system/globalid/lnuid.h
+++ b/system/globalid/lnuid.h
@@ -1,0 +1,53 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2019 HPCC Systems®.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+############################################################################## */
+
+#include <ctime>
+#include <string>
+
+using namespace std;
+
+/**
+* This is an implementation of Globally Unique Transaction ID’s.
+* Note: This class currently generates unique id on OS's that supports device /dev/urandom.
+* @author - Amol Patwardhan
+* Version: 1.0
+*/
+
+namespace ln_uid {
+
+    //Random byte count
+    const unsigned int uid_size = 16;
+
+    typedef unsigned char ln_uid_t[uid_size];
+
+    ln_uid_t &createUniqueId(ln_uid_t &out);
+    string createUniqueIdString();
+
+    std::string uniqueIdToString(const ln_uid_t &uid);
+    ln_uid_t &uniqueIdFromString(const char* uid, ln_uid_t &out);
+
+    time_t timeFromUniqueId(const char* uid);
+    time_t timeFromUniqueId(const ln_uid_t &uid);
+
+    void getUniqueIdRange(time_t start, time_t end, ln_uid_t &uid_start, ln_uid_t &uid_end);
+    void getUniqueIdDateRange(const char *start, const char *end, ln_uid_t &uid_start, ln_uid_t &uid_end);
+
+    bool sameUniqueId(const ln_uid_t &uid1, const ln_uid_t &uid2);
+    void copyUniqueId(ln_uid_t &to, const ln_uid_t &from);
+
+    int get_utc_offset();
+};

--- a/system/jlib/CMakeLists.txt
+++ b/system/jlib/CMakeLists.txt
@@ -92,6 +92,7 @@ set (    SRCS
          jtime.cpp 
          junicode.cpp 
          jutil.cpp 
+         ${HPCC_SOURCE_DIR}/system/globalid/lnuid.cpp
          sourcedoc.xml
     )
 
@@ -188,7 +189,7 @@ include_directories (
          ${HPCC_SOURCE_DIR}/system/include
          ${HPCC_SOURCE_DIR}/system/lzma
          ${HPCC_SOURCE_DIR}/system/lz4_sm/lz4/lib
-         ${HPCC_SOURCE_DIR}/system/libbase58
+         ${HPCC_SOURCE_DIR}/system/globalid
          ${HPCC_SOURCE_DIR}/system/security/cryptohelper
          ${CMAKE_CURRENT_BINARY_DIR}  # for generated jelog.h file 
          ${CMAKE_BINARY_DIR}

--- a/system/jlib/jlog.cpp
+++ b/system/jlib/jlog.cpp
@@ -28,8 +28,9 @@
 #include "jsocket.hpp"
 #include "jmisc.hpp"
 #include "jprop.hpp"
+#include "lnuid.h"
 
-#include "libbase58.h"
+using namespace ln_uid;
 
 #define MSGCOMP_NUMBER 1000
 #define FILE_LOG_ENABLES_QUEUEUING
@@ -2580,7 +2581,7 @@ public:
     virtual void setGlobalId(const char *id, SocketEndpoint &ep, unsigned pid) override
     {
         globalId.set(id);
-        appendLocalId(localId.clear(), ep, pid);
+        appendGloballyUniqueId(localId.clear());
     }
     virtual void setCallerId(const char *id) override
     {
@@ -2625,28 +2626,11 @@ extern jlib_decl IContextLogger &updateDummyContextLogger()
     return dummyContextLogger;
 }
 
-extern jlib_decl StringBuffer &appendLocalId(StringBuffer &s, const SocketEndpoint &ep, unsigned pid)
+
+extern jlib_decl StringBuffer &appendGloballyUniqueId(StringBuffer &s)
 {
-    static unsigned short cnt = msTick();
-
-    MemoryBuffer data;
-    data.append(ep.iphash());
-    if (pid>0)
-        data.append(pid);
-    else
-        data.append(ep.port);
-    data.append(++cnt);
-
-    size_t b58Length = data.length() * 2;
-    StringBuffer id;
-
-    //base58 works well as a human readable format, i.e. so customers can easily report the id that was recieved
-    if (b58enc(id.reserve(b58Length), &b58Length, data.toByteArray(), data.length()) && b58Length > 1)
-    {
-        id.setLength(b58Length);
-        s.append(id);
-    }
-    return s;
+    string uid = createUniqueIdString();
+    return s.append(uid.c_str());
 }
 
 extern jlib_decl void UseSysLogForOperatorMessages(bool use)

--- a/system/jlib/jlog.hpp
+++ b/system/jlib/jlog.hpp
@@ -1002,7 +1002,7 @@ interface jlib_decl IContextLogger : extends IInterface
     virtual const char *queryCallerId() const = 0;
 };
 
-extern jlib_decl StringBuffer &appendLocalId(StringBuffer &s, const SocketEndpoint &ep, unsigned pid);
+extern jlib_decl StringBuffer &appendGloballyUniqueId(StringBuffer &s);
 
 extern jlib_decl const IContextLogger &queryDummyContextLogger();
 extern jlib_decl IContextLogger &updateDummyContextLogger();

--- a/testing/regress/ecl/soapcall.ecl
+++ b/testing/regress/ecl/soapcall.ecl
@@ -20,6 +20,8 @@
 
 //version targetIP='127.0.0.1',goodPort='9876',blacListedPort='9875'
 
+#option('generateGlobalId', true);
+
 string targetIP := '.' : stored('targetIP');
 string goodPort := '9876' : stored('goodPort');
 string blacListedPort := '9875' : stored('blacListedPort');

--- a/thorlcr/graph/thgraph.cpp
+++ b/thorlcr/graph/thgraph.cpp
@@ -2568,7 +2568,7 @@ public:
     virtual void setGlobalId(const char *id, SocketEndpoint &ep, unsigned pid) override
     {
         globalId.set(id);
-        appendLocalId(localId.clear(), ep, pid);
+        appendGloballyUniqueId(localId.clear());
     }
     virtual void setCallerId(const char *id) override
     {

